src/htm/algorithms/Connections.cpp

```diff
@@ -654,6 +654,16 @@ void Connections::bumpSegment(const Segment segment, const Permanence delta) {
 }
 
 
+vector<CellIdx> Connections::presynapticCellsForSegment(const Segment segment) const { //TODO optimize by storing the vector in SegmentData?
```

Comment:
this is the new method reimplemented in c++. 
@fcr this is an old PR, what is a heavy user of the functionality, how do I compare the speed to the original py version?

Reply:
@breznak I'll need to refresh my memory. 
I am tied up at the moment but I'll be free to look at it on Thursday.

Reply:
thank you very much, no hurry. 
This was about speeding up access to presynaptic cells via `presynapticCellsForSegment()` which you've written/ported to htm.advanced, and we wanted to speedup with a direct cpp implementation (so far the naive implementation is imho same/bit slower).

Reply:
@breznak the attached quick and dirty test times the growSynapseToSample() using a Thalamus instance.

The C++ version results in 3x speedup, so well done!
I've also attached a screen shot of the profiles before and after the C++ version
<img width="2633" alt="Screen Shot 2020-09-10 at 15 12 29" src="https://user-images.githubusercontent.com/103160/92728339-38566580-f379-11ea-9486-8c8664b58abc.png">

from htm.advanced.algorithms.thalamus import Thalamus
from htm.bindings.sdr import SDR
import numpy as np
import time

class ThalamusTimingTest():

    def _inferThalamus(self, t, l6Input, ffInput):
        """
        Compute the effect of this feed forward input given the specific L6 input.
    
        :param t: instance of Thalamus
        :param l6Input:
        :param ffInput: a numpy array of 0's and 1's
        :return:
        """
        t.reset()
        t.deInactivateCells(l6Input)
        ffOutput = t.computeFeedForwardActivity(ffInput)
        return ffOutput
      
    # Simple tests for debugging
    def _trainThalamus(self, t):
        # Learn
        L6Pattern = SDR(t.l6CellCount)
        L6Pattern.sparse = [0, 1, 2, 3, 4, 5]
        t.learnL6Pattern(L6Pattern, [(0, 0), (2, 3)])
        L6Pattern.sparse = [6, 7, 8, 9, 10]
        t.learnL6Pattern(L6Pattern, [(1, 1), (3, 4)])   
    
    def run(self, verbose=False):
        """
        Test that thalamus relays around the trained locations,
        and also does busrts.
        """
        t = Thalamus(trnThreshold=6)
    
        self._trainThalamus(t)
        ff = np.zeros((32,32))
        ff.reshape(-1)[[8, 9, 98, 99]] = 1.0
        
        L6Pattern = SDR(t.l6CellCount)
        L6Pattern.sparse = [0, 1, 2, 3, 4, 5]
        
        self._inferThalamus(t, L6Pattern, ff)
        
if __name__ == '__main__':
    test = ThalamusTimingTest()
    
    start= time.time()
    for _ in range(100):
        test.run()
    end = time.time()
    print("total time = {:3.2f} sec".format(end - start))

Reply:
Glad to hear! Thank you for the microbenchmark and confirming. 
@fcr can you evaluate how useful the function is (for you and) in general for research purposes? For a relatively small overhead, I can store the data in each Segment, and then the query for presynaptic cells would be costless `O(1)`

Reply:
Awesome, 3x times faster is nice,
will this improvement also make faster `presynapticCellForSynapse()` ? I am using that in HTMPandaVis

Reply:
@breznak I think that @Zbysekz just gave an answer to that :-)

Reply:
> will this improvement also make faster presynapticCellForSynapse() ? I am using that in HTMPandaVis 

@Zbysekz my post above addresses that - I don't think it'll be faster now. We can make it instant with slight overhead for everybody in `updateSynapsePermanence()`. 

I suggest we merge this as "grow synapses now faster", and I'll look at the presynapticCells in a followup. 
Zbysek, can you show me the code where you're using the presyns? Just to get an idea.

Reply:
@breznak code is there: [https://github.com/htm-community/HTMpandaVis/blob/b2c81ad3c3eac7aa0edff13c6fa170aa1dbf26fe/PandaVis/bakeReader/bakeReader.py#L201](https://github.com/htm-community/HTMpandaVis/blob/b2c81ad3c3eac7aa0edff13c6fa170aa1dbf26fe/PandaVis/bakeReader/bakeReader.py#L201)